securedrop_client/gui/widgets.py

```diff
@@ -875,23 +875,57 @@ def update(self, sources: List[Source]) -> List[str]:
         # also clear the source_widgets dict.
         self.clear()
         self.source_widgets = {}
+        sources.reverse()
+        self.add_source(existing_source_widgets, sources, current_source_id,
+                        has_source_selected)
 
-        for source in sources:
-            new_source = SourceWidget(source)
-            new_source.setup(self.controller)
-            self.source_widgets[source.uuid] = new_source
-
-            list_item = QListWidgetItem(self)
-            list_item.setSizeHint(new_source.sizeHint())
-
-            self.addItem(list_item)
-            self.setItemWidget(list_item, new_source)
+    def add_source(self, existing_source_widgets, sources, current_source_id,
+                   has_source_selected, slice_size=1):
+        """
+        Add a slice of sources, ensure the currently selected source is
+        retained and, if necessary, reschedule the addition of more sources.
 
-            if source.id == current_source_id and has_source_selected:
-                self.setCurrentItem(list_item)
+        If no more sources are left, work out the deleted sources and emit
+        deletion signal for each source to delete.
+        """
 
-        deleted_uuids = list(set(existing_source_widgets.keys()) - set(self.source_widgets.keys()))
-        return deleted_uuids
+        def schedule_source_management(slice_size=slice_size):
+            if not sources:
+                # No more sources to add, discover the deleted sources and
+                # delete them.
+                deleted_uuids = list(set(existing_source_widgets.keys()) -
+                                     set(self.source_widgets.keys()))
+                for source_uuid in deleted_uuids:
+                    self.delete_source_by_uuid.emit(source_uuid)
```

Comment:
well this is interesting, I'm reliably getting a crash when I try to delete a source locally:

```
2020-03-10 10:27:05,208 - root:53(excepthook) ERROR: Unrecoverable error
Traceback (most recent call last):
  File "/Users/redshiftzero/Documents/Github/securedrop-client/securedrop_client/gui/widgets.py", line 730, in delete_source
    conversation_wrapper.deleteLater()
RuntimeError: wrapped C/C++ object of type SourceConversationWrapper has been deleted
```

investigating...

Reply:
It'll be something I've missed. Sorry about that. Perhaps drop this for now and let me investigate tomorrow and fix..?

Reply:
I did a bit of debugging and can't yet see where the `SourceConversationWrapper` is getting deleted 🤔 but if someone figures it out today (we'll timebox) we'll post a note on here.

Reply:
This is, ultimately, what's called: https://github.com/freedomofpress/securedrop-client/blob/55f3dd8467f10f6573f9eac91a675a3670e975ae/securedrop_client/gui/widgets.py#L722

Reply:
yep, but the underlying C++ object corresponding to the Python wrapper is getting deleted by qt _before_ we run `conversation_wrapper.deleteLater` in that method (that's what's causing the `RuntimeError` above)

Reply:
:+1: sure, I get that... :slightly_smiling_face: I was just filling in the gaps. I'll take a look tomorrow.

Reply:
OK. Through the ubiquitous use of the `print` statement, I think I've found the problem and provided a fix.

As this branch currently stands, when a source is deleted two things happen in the following order:

1. The `on_source_changed` method is called. If no source is selected (and this is the case, since we've just deleted it), then `clear_conversation` is called. This ultimately calls `child.widget().deleteLater()` (where `child` is the conversation wrapper.
2. Meanwhile, the adding of current sources and deletion of remaining sources is scheduled in the future (as per the strategy used in the PR). It is when the `delete_source_by_uuid` signal was emitted that we hit the problem code where the `RuntimeError` was raised. Put simply, the underlying widget had been earlier GC'd (see 1) but the Python object was still in the `source_conversations` dictionary.

The (simple) solution is to not call `conversation_wrapper.deleteLater()` at that later scheduled time since we'd already deleted the the underlying widget with the call in 1. The conversation_wrapper, by being removed from the dict also goes out of scope and is GC'd.

I hope this is clear...? :grimacing: :wink: